tests/test-json/src/main.rs

```diff
@@ -67,102 +67,142 @@ mod tests {
-
     // Maximum duration to wait for DTrace, controlling total test duration
     const MAX_WAIT: Duration = Duration::from_secs(30);
 
-    fn run_test(recv: Receiver<()>) {
+    // A sentinel printed by DTrace, so we know when it starts up successfully.
+    const BEGIN_SENTINEL: &str = "BEGIN";
+
+    // Fire the test probes in sequence, when a notification is received on the channel.
+    async fn fire_test_probes(mut recv: Receiver<()>) {
         usdt::register_probes().unwrap();
 
         // Wait for notification from the main thread
-        recv.recv().unwrap();
-        sleep(SLEEP_DURATION);
+        println!("Test runner waiting for first notification");
+        recv.recv().await.unwrap();
         println!("Test runner firing first probe");
 
         // Fire the good probe until the main thread signals us to continue.
         let data = ProbeArg::default();
         test_json::good!(|| &data);
-        println!("Test runner awaiting notification");
-        recv.recv().unwrap();
+        println!("Test runner fired first probe");
+        println!("Test runner awaiting notification to continue");
+        recv.recv().await.unwrap();
+        println!("Test runner received notification to continue");
 
         // Fire the bad probe.
-        sleep(SLEEP_DURATION);
         println!("Test runner firing second probe");
         let data = NotJsonSerializable::default();
         test_json::bad!(|| &data);
+        println!("Test runner fired second probe");
+    }
+
+    // Run DTrace as a subprocess, waiting for the JSON output of the provided probe.
+    async fn run_dtrace_and_return_json(tx: &Sender<()>, probe_name: &str) -> Value {
+        // Start the DTrace subprocess, and don't exit if the probe doesn't exist.
+        let mut dtrace = Command::new(root_command())
+            .arg("dtrace")
+            .arg("-Z")
+            .arg("-q")
+            .arg("-n")
+            // The test probe we're interested in listening for.
+            .arg(format!(
+                "test_json{}:::{} {{ printf(\"%s\", copyinstr(arg0)); exit(0); }}",
+                std::process::id(),
+                probe_name
+            ))
+            .arg("-n")
+            // An output printed by DTrace when it starts, to coordinate with the test thread
+            // firing the probe itself.
+            .arg(format!("BEGIN {{ trace(\"{}\"); }}", BEGIN_SENTINEL))
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .spawn()
+            .expect("Failed to spawn DTrace subprocess");
+
+        // Wait for DTrace to correctly start up before notifying the test thread to start
+        // firing probes.
+        let now = Instant::now();
+        wait_for_begin_sentinel(&mut dtrace, &now).await;
+
+        // We should now see the probe having fired exactly once. Grab the output data, kill the
+        // DTrace process, and then parse the output as JSON.
+        tx.send(()).await.unwrap();
```

Comment:
I don't feel like this comment goes with this code... although I think this is carried over from the previous version.

Reply:
Ah, yeah I meant to remove that, thanks for catching.